### PR TITLE
MAINT: Improve performance of np.full

### LINKED ISSUE #16180: full works on lists but with very bad performance
When you pass a python list to full, it actually works as expected but with bad performance. The documentation says that only a scalar is accepted, but because it does the right thing, it's very easy for people to end up with code like this that's correct but inefficient.

```
import numpy
x = numpy.full(100, range(100))
```

The reason it's inefficient is because the source looks like this:
```
    if dtype is None:
        dtype = array(fill_value).dtype
    a = empty(shape, dtype, order)
    multiarray.copyto(a, fill_value, casting='unsafe')
    return a

```
If the dtype is not specified, the branch actually creates a copy of the whole array, throws it away, and then does it again.

I would suggest either explicitly supporting this use case, and improving performance and docs to match. Or, try to do some kind of basic check to reject this input as unacceptable.

### PR DESCRIPTION
**Pull request by ghost**
Implement more performant np.full and add test

Closes #16180

### CODE REVIEW

**Comment by vrakesh** [comment]
Thanks for the fix @dojafrat , requesting to add some basic speedup numbers here. This would serve as a reference for the future.

File: numpy/core/tests/test_api.py [hunk 1]
```diff
@@ -542,3 +542,10 @@ def test_broadcast_arrays():
     result = np.broadcast_arrays(a, b)
     assert_equal(result[0], np.array([(1, 2, 3), (1, 2, 3), (1, 2, 3)], dtype='u4,u4,u4'))
     assert_equal(result[1], np.array([(1, 2, 3), (4, 5, 6), (7, 8, 9)], dtype='u4,u4,u4'))
+
+@pytest.mark.parametrize(["shape", "fill_value", "expected_output"],
+        [((2, 2), [5.0,  6.0], np.array([[5.0, 6.0], [5.0, 6.0]])),
+         ((3, 2), [1.0,  2.0], np.array([[1.0, 2.0], [1.0, 2.0], [1.0,  2.0]]))])
+def test_full_from_list(shape, fill_value, expected_output):
+    output = np.full(shape, fill_value)
+    assert_equal(output, expected_output)
```
Comment (eric-wieser) [inline numpy/core/tests/test_api.py:551]:
```suggestion
    assert_equal(output, expected_output)

```

**Review by eric-wieser — APPROVED** [review]
As discussed in the original issue / PR, I see no point measuring the performance of this, the improvement is both an optimization _and_ a simplification.

**Comment by seberg** [comment]
Sorry for not merging earlier. Thanks @dojafrat and Eric.

**Pull request merged**